src/Microsoft.Framework.PackageManager/Restore/NuGet/NuGetv3Feed.cs

```diff
@@ -0,0 +1,298 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading.Tasks;
+using NuGet;
+using Newtonsoft.Json.Linq;
+using Newtonsoft.Json;
+using Microsoft.Framework.Runtime.Internal;
+
+namespace Microsoft.Framework.PackageManager.Restore.NuGet
+{
+    internal class NuGetv3Feed : IPackageFeed
+    {
+        private readonly string _baseUri;
+        private readonly Reports _reports;
+        private readonly HttpSource _httpSource;
+        private readonly TimeSpan _cacheAgeLimitList;
+        private readonly TimeSpan _cacheAgeLimitNupkg;
+        private readonly bool _ignoreFailure;
+        private bool _ignored;
+
+        private readonly Dictionary<string, Task<IEnumerable<PackageInfo>>> _packageVersionsCache = new Dictionary<string, Task<IEnumerable<PackageInfo>>>();
+        private readonly Dictionary<string, Task<NupkgEntry>> _nupkgCache = new Dictionary<string, Task<NupkgEntry>>();
+
+        public string Source
+        {
+            get
+            {
+                return _httpSource.BaseUri;
+            }
+        }
+
+        internal NuGetv3Feed(
+            HttpSource httpSource,
+            bool noCache,
+            Reports reports,
+            bool ignoreFailure)
+        {
+            var baseUri = httpSource.BaseUri;
+            _baseUri = baseUri.EndsWith("/index.json") ? baseUri.Substring(0, baseUri.Length - "index.json".Length) : baseUri;
+            _reports = reports;
+            _httpSource = httpSource;
+            _ignoreFailure = ignoreFailure;
+            if (noCache)
+            {
+                _cacheAgeLimitList = TimeSpan.Zero;
+                _cacheAgeLimitNupkg = TimeSpan.Zero;
+            }
+            else
+            {
+                _cacheAgeLimitList = TimeSpan.FromMinutes(30);
+                _cacheAgeLimitNupkg = TimeSpan.FromHours(24);
+            }
+        }
+
+        internal static bool DetectNuGetV3(HttpSource httpSource, out Uri packageBaseAddress)
+        {
+            try
+            {
+                var result = httpSource.GetAsync(httpSource.BaseUri, "index_json", TimeSpan.FromHours(6)).Result;
+                using (var reader = new StreamReader(result.Stream))
+                {
+                    var content = reader.ReadToEnd();
+                    var indexJson = JObject.Parse(content);
```

Comment:
You could load the JObject directly from the StreamReader and avoid allocating the string. You're already doing this on line 141.